File: examples/mobile/android/MobileBenchmarkUploader.kt
package ai.tollama.splineforecast.mobile

import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

object MobileBenchmarkUploader {
    private const val HMAC_ALGORITHM = "HmacSHA256"

    fun upload(
        endpoint: String,
        apiToken: String?,
        idempotencyKey: String?,
        runId: String,
        deviceProfile: String,
        expectedPlatform: String,
        benchmarkPayloadJson: String,
        signingSecret: String? = null,
        batch: Boolean = false,
    ): Int {
        val requestJson = if (batch) {
            """
                {
                  "uploads": [
                    {
                      "run_id": "$runId",
                      "device_profile": "$deviceProfile",
                      "expected_platform": "$expectedPlatform",
                      "benchmark_result": $benchmarkPayloadJson
                    }
                  ]
                }
            """.trimIndent()
        } else {
            """
                {
                  "run_id": "$runId",
                  "device_profile": "$deviceProfile",
                  "expected_platform": "$expectedPlatform",
                  "benchmark_result": $benchmarkPayloadJson
                }
            """.trimIndent()
        }

        val target = if (batch && endpoint.endsWith(":ingest")) {
            endpoint.removeSuffix(":ingest") + ":ingest-batch"
        } else {
            endpoint
        }
        val conn = URL(target).openConnection() as HttpURLConnection
        conn.requestMethod = "POST"
        conn.doOutput = true
        conn.setRequestProperty("Content-Type", "application/json")
        apiToken?.let { conn.setRequestProperty("X-API-Token", it) }
        idempotencyKey?.let { conn.setRequestProperty("X-Idempotency-Key", it) }
        if (!signingSecret.isNullOrBlank()) {
            val timestamp = (System.currentTimeMillis() / 1000L).toString()
            val signature = hmacSha256Hex(signingSecret, "$timestamp\n$requestJson")
            conn.setRequestProperty("X-Mobile-Timestamp", timestamp)
            conn.setRequestProperty("X-Mobile-Signature", signature)
        }

        OutputStreamWriter(conn.outputStream, Charsets.UTF_8).use { it.write(requestJson) }
        return conn.responseCode
    }

    private fun hmacSha256Hex(secret: String, message: String): String {
        val key = SecretKeySpec(secret.toByteArray(Charsets.UTF_8), HMAC_ALGORITHM)
        val mac = Mac.getInstance(HMAC_ALGORITHM)
        mac.init(key)
        return mac.doFinal(message.toByteArray(Charsets.UTF_8)).joinToString("") { byte -> "%02x".format(byte) }
    }
}

File: examples/mobile/ios/MobileBenchmarkUploader.swift
import Foundation
import CryptoKit

enum MobileBenchmarkUploader {
    static func makeRequest(
        endpoint: URL,
        apiToken: String?,
        idempotencyKey: String?,
        runId: String,
        deviceProfile: String,
        expectedPlatform: String,
        benchmarkPayload: [String: Any],
        signingSecret: String? = nil,
        batch: Bool = false
    ) throws -> URLRequest {
        let body: [String: Any]
        if batch {
            body = [
                "uploads": [[
                    "run_id": runId,
                    "device_profile": deviceProfile,
                    "expected_platform": expectedPlatform,
                    "benchmark_result": benchmarkPayload,
                ]],
            ]
        } else {
            body = [
                "run_id": runId,
                "device_profile": deviceProfile,
                "expected_platform": expectedPlatform,
                "benchmark_result": benchmarkPayload,
            ]
        }
        let data = try JSONSerialization.data(withJSONObject: body, options: [.prettyPrinted, .sortedKeys])
        let target: URL
        if batch, endpoint.absoluteString.hasSuffix(":ingest"), let url = URL(string: endpoint.absoluteString.replacingOccurrences(of: ":ingest", with: ":ingest-batch")) {
            target = url
        } else {
            target = endpoint
        }
        var request = URLRequest(url: target)
        request.httpMethod = "POST"
        request.httpBody = data
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let apiToken {
            request.setValue(apiToken, forHTTPHeaderField: "X-API-Token")
        }
        if let idempotencyKey {
            request.setValue(idempotencyKey, forHTTPHeaderField: "X-Idempotency-Key")
        }
        if let signingSecret, !signingSecret.isEmpty {
            let timestamp = String(Int(Date().timeIntervalSince1970))
            let signature = hmacSha256Hex(secret: signingSecret, message: "\(timestamp)\n" + String(decoding: data, as: UTF8.self))
            request.setValue(timestamp, forHTTPHeaderField: "X-Mobile-Timestamp")
            request.setValue(signature, forHTTPHeaderField: "X-Mobile-Signature")
        }
        return request
    }

    private static func hmacSha256Hex(secret: String, message: String) -> String {
        let key = SymmetricKey(data: Data(secret.utf8))
        let digest = HMAC<SHA256>.authenticationCode(for: Data(message.utf8), using: key)
        return digest.map { String(format: "%02x", $0) }.joined()
    }
}
